### PR TITLE
#70

### PR DESCRIPTION
**Pull request by motoyasu-saburi**
# Details
https://github.com/apiaryio/dredd/issues/677
https://github.com/apiaryio/dredd-transactions/issues/70

Please give me your opinion on terms and test cases.

### CODE REVIEW

**Comment by motoyasu-saburi** [comment]
oh, sry.
⬆️ this error cause is my commit message.

**Comment by honzajavorek** [comment]
@motoyasu-saburi I'm sorry for the delay! (I've got sick.) Thank you very much for the contribution, I'm pretty excited to see what we have here. Going to look at it right away 🙂 🚀

File: test/unit/expand-uri-template-with-parameters-test.coffee [hunk 1]
```diff
@@ -355,3 +355,48 @@ describe 'expandUriTemplateWithParameters', ->
 
           it 'should return some URI', ->
             assert.isNotNull data['uri']
+
+
+        describe 'when example value are not given', ->
```
Comment (honzajavorek) [inline test/unit/expand-uri-template-with-parameters-test.coffee]: Typo:

```diff
- when example value are not given
+ when example value is not given
```

File: test/unit/expand-uri-template-with-parameters-test.coffee [hunk 2]
```diff
@@ -355,3 +355,48 @@ describe 'expandUriTemplateWithParameters', ->
...
+          it 'should return no error', ->
+            assert.equal data['errors'].length, 0
+
+          it 'should return 1 warning', ->
+            assert.equal data['warnings'].length, 1
+
+          it 'should return some URI', ->
+            assert.isNotNull data['uri']
+
+        describe 'when default values are not given', ->
```
Comment (honzajavorek) [inline test/unit/expand-uri-template-with-parameters-test.coffee]: Typo:

```diff
- when default values are not given
+ when default value is not given
```

File: test/unit/validate-parameters-test.coffee [hunk 3]
```diff
@@ -248,3 +248,31 @@ describe 'validateParameters', ->
...
+          name:
+            description: 'Machine name'
+            type: 'boolean'
+            required: true
+            example: ''
+            default: 'false'
+            values: []
+
+        result = validateParameters params
+        assert.equal result['errors'].length, 1
```
Comment (honzajavorek) [inline test/unit/validate-parameters-test.coffee]: This test says "should not set the error", but then there's

```coffee
assert.equal result['errors'].length, 1
```

which asserts the opposite.

**Comment by honzajavorek** [comment]
@motoyasu-saburi @kumagoro-obscura Any change you would get to this? It should be just a small change to allow empty string as a valid value for the parameters.

**Comment by motoyasu-saburi** [comment]
@honzajavorek 
Sorry to have taken time.
I will reply to 🔼 by looking at this week. 🙇

**Comment by honzajavorek** [comment]
@motoyasu-saburi No problem! Looking forward 👍

**Comment by motoyasu-saburi** [comment]
Thank you for your explanation :)
```
One thing I'd like to discuss is the behavior when example or default values are empty strings. In my opinion, the correct behavior would be to generate /honey?param= instead of excluding the parameter. What do you think?
```
I certainly thought it is correct. 

I'm going to fix conditions and test cases, commit message. 👍

**Comment by motoyasu-saburi** [comment]
I would like to ask you the following questions.
This test case describe is `when parameter is required and example and default value are empty`, but I thought `example` and `default` is `empty string`.
Why the parameters isn't null?
https://github.com/apiaryio/dredd-transactions/blob/master/test/unit/validate-parameters-test.coffee#L163

**Comment by honzajavorek** [comment]
@motoyasu-saburi The reason why the current test says `empty` and the values are `empty string` is that when the original author of the test wrote it, they didn't realize `empty` can mean many different things. So they wrote `empty` and tested `empty string`, while `null` or `undefined` etc. was left untested.

But the test brings up a good question. If empty string should be considered to be a valid value, then this test needs to be changed. We would just need to verify that if user specifies no example/default value in the API description, it's propagated as `null` or `undefined` to the conditions we're fixing here.

The problem I see and which I did not realize earlier - there's probably no easy way how to specify an empty string default/example value for an URI parameter in the API Blueprint:

image

@kylef @pksunkara Is "empty string as a value" supported scenario in the API Blueprint?

**Comment by pksunkara** [comment]
```
+ id (string)
```

has an empty value.

**Comment by honzajavorek** [comment]
@pksunkara Good point, it's vice versa. Basically, all URI parameters in API Blueprint default to having empty string as a value unless they're specified as `nullable`, is that correct?

**Comment by motoyasu-saburi** [comment]
@honzajavorek 
I need to do the following, right?

* Correct the test case of `empty string`
* Add a `null` pattern of test case.

**Comment by honzajavorek** [comment]
@motoyasu-saburi Yes, that sounds like a good plan 👍

**Comment by honzajavorek** [comment]
Closing in favor of https://github.com/apiaryio/dredd-transactions/pull/72

**Pull request closed without merging**